src/librustc_mir/transform/const_prop.rs

```diff
@@ -636,18 +643,40 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     ScalarMaybeUndef::Scalar(one),
                     ScalarMaybeUndef::Scalar(two)
                 ) => {
-                    let ty = &value.layout.ty.kind;
+                    // Value represented as a pair, but we still want to generate zsts for rest of
+                    // the fields.
+                    let layout = &value.layout;
+                    debug!("replace_with_const: value layout={:#?}", layout);
+                    debug!("replace_with_const: value field count={}", layout.fields.count());
+                    let ty = &layout.ty.kind;
                     if let ty::Tuple(substs) = ty {
+                        // Values from the pair will be popped from this vec when we see a non-zst
+                        // type in the layout
+                        let mut pair_values = vec![two, one];
+                        // Final values in the aggregate
+                        let mut aggregate_values = vec![];
+                        // Iterate fields, for zsts a zsg scalar, for others use the values from
+                        // the pair
+                        for i in 0 .. layout.fields.count() {
+                            let field_layout = layout.field(self, i).unwrap();
+                            debug!("field {} layout={:#?}", i, field_layout);
+                            let scalar = if field_layout.details.size == Size::ZERO {
+                                Scalar::zst()
+                            } else {
+                                pair_values.pop().unwrap()
```

Comment:
I think there might be a need for an additional check that the field actually
has `Abi::Scalar` layout. Otherwise it could end up with an assignment between
`Scalar` and `ScalarPair` layouts, for example in something like `test(((1, 2),))`,
when propagating the value of the outermost tuple.

Reply:
Amazing, your example uncovers another bug. I reported it as #67019.

I'll try to fix it in this PR too.